ambry-router/src/main/java/com.github.ambry.router/DefaultGCMCryptoService.java

```diff
@@ -0,0 +1,298 @@
+ * This implementation uses GCM for encryption and decryption
+ */
+class DefaultGCMCryptoService implements CryptoService<SecretKeySpec> {
+
+  private static final int Version_Field_Size_In_Bytes = Short.BYTES;
+  private static final short KeyRecord_Version_V1 = 1;
+  public static final short IVRecord_Version_V1 = 1;
+  private static final String GCM_CRYPTO_INSTANCE = "AES/GCM/NoPadding";
+  private static final SecureRandom random = new SecureRandom();
+  private final int ivValSize;
+  private final int keySize;
+  private final CryptoServiceConfig config;
+
+  private static final Logger logger = LoggerFactory.getLogger(DefaultGCMCryptoService.class);
+
+  DefaultGCMCryptoService(CryptoServiceConfig cryptoServiceConfig) {
+    config = cryptoServiceConfig;
+    ivValSize = cryptoServiceConfig.cryptoServiceIvSizeInBytes;
+    keySize = cryptoServiceConfig.cryptoServiceKeySizeInChars;
+    Security.addProvider(new BouncyCastleProvider());
+    if (!config.cryptoServiceEncryptionDecryptionMode.equals("GCM")) {
+      throw new IllegalArgumentException(
+          "Unrecognized Encryption Decryption Mode " + config.cryptoServiceEncryptionDecryptionMode);
+    }
+  }
+
+  @Override
+  public ByteBuffer encrypt(ByteBuffer toEncrypt, SecretKeySpec key) throws GeneralSecurityException {
+    try {
+      Cipher encrypter = Cipher.getInstance(GCM_CRYPTO_INSTANCE, "BC");
+      byte[] iv = new byte[ivValSize];
+      random.nextBytes(iv);
+      encrypter.init(Cipher.ENCRYPT_MODE, key, new IvParameterSpec(iv));
+      int outputSize = encrypter.getOutputSize(toEncrypt.remaining());
+      ByteBuffer encryptedContent = ByteBuffer.allocate(IVRecord_Format_V1.getIVRecordSize(iv) + outputSize);
+      IVRecord_Format_V1.serializeIVRecord(encryptedContent, iv);
+      encrypter.doFinal(toEncrypt, encryptedContent);
+      encryptedContent.flip();
+      return encryptedContent;
+    } catch (Exception e) {
+      throw new GeneralSecurityException("Exception thrown while encrypting data", e);
+    }
+  }
+
+  @Override
+  public ByteBuffer decrypt(ByteBuffer toDecrypt, SecretKeySpec key) throws GeneralSecurityException {
+    try {
+      Cipher decrypter = Cipher.getInstance(GCM_CRYPTO_INSTANCE, "BC");
+      byte[] iv = deserializeIV(new ByteBufferInputStream(toDecrypt));
+      decrypter.init(Cipher.DECRYPT_MODE, key, new IvParameterSpec(iv));
+      ByteBuffer decryptedContent = ByteBuffer.allocate(decrypter.getOutputSize(toDecrypt.remaining()));
+      decrypter.doFinal(toDecrypt, decryptedContent);
+      decryptedContent.flip();
+      return decryptedContent;
+    } catch (Exception e) {
+      throw new GeneralSecurityException("Exception thrown while decrypting data", e);
+    }
+  }
+
+  @Override
+  public SecretKeySpec getRandomKey() {
+    return new SecretKeySpec(Hex.decode(CryptoUtils.getRandomKey(keySize, random)), config.cryptoServiceKeyGenAlgo);
```

Comment:
can `CryptoUtils.getRandomKey` return a `byte[]` directly instead of converting to hex and back to raw bytes?